src/main/java/org/spongepowered/common/data/processor/data/entity/GameModeDataProcessor.java

```diff
@@ -0,0 +1,176 @@
+import org.spongepowered.api.service.persistence.InvalidDataException;
+import org.spongepowered.common.Sponge;
+import org.spongepowered.common.data.ImmutableDataCachingUtil;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeGameModeData;
+import org.spongepowered.common.data.processor.common.AbstractSpongeDataProcessor;
+import org.spongepowered.common.data.util.DataUtil;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+public class GameModeDataProcessor extends AbstractSpongeDataProcessor<GameModeData, ImmutableGameModeData> {
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof Player;
+    }
+
+    @Override
+    public Optional<GameModeData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            if (dataHolder instanceof EntityPlayerMP) {
+                Optional.of(new SpongeGameModeData(
+                        (GameMode) ((Object) ((EntityPlayerMP) dataHolder).theItemInWorldManager.getGameType())));
+            }
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<GameModeData> fill(DataHolder dataHolder, GameModeData manipulator) {
+        if (dataHolder instanceof EntityPlayerMP) {
+            return Optional.of(manipulator.set(Keys.GAME_MODE,
+                    (GameMode) ((Object) ((EntityPlayerMP) dataHolder).theItemInWorldManager.getGameType())));
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<GameModeData> fill(DataHolder dataHolder, GameModeData manipulator, MergeFunction overlap) {
+        if (dataHolder instanceof EntityPlayerMP) {
+            final GameModeData original = from(dataHolder).get();
+            return Optional.of(manipulator.set(Keys.GAME_MODE, overlap.merge(manipulator, original).type().get()));
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<GameModeData> fill(DataContainer container, GameModeData gameModeData) {
+        final String modeId = DataUtil.getData(container, Keys.GAME_MODE, String.class);
+        final Optional<GameMode> optional = Sponge.getSpongeRegistry().getType(GameMode.class, modeId);
+        if (optional.isPresent()) {
+            gameModeData.set(Keys.GAME_MODE, optional.get());
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, GameModeData manipulator) {
+        if (dataHolder instanceof EntityPlayerMP) {
+            final GameMode oldMode =
+                    (GameMode) ((Object) ((EntityPlayerMP) dataHolder).theItemInWorldManager.getGameType());
+            final ImmutableValue<GameMode> newMode = manipulator.type().asImmutable();
+            try {
+                ((EntityPlayerMP) dataHolder).setGameType((WorldSettings.GameType) ((Object) manipulator.type().get()));
+                return DataTransactionBuilder.builder()
```

Comment:
You could simplify this with `DataTransactionBuilder.successReplace(ImmutableDataCachingUtil...., newMode);`